Add support for new resource and datasource for Managed DevOps Pools service (#28325)
terraform-providers/terraform-provider-azurerm: website/docs/r/managed_devops_pool.html.markdown

* `dev_center_project_resource_id` - (Required) The ID of the Dev Center project.

* `fabric_profile` - (Required) A `fabric_profile` block as defined below.

* `location` - (Required) The Azure Region where the Managed DevOps Pool should exist. Changing this forces a new Managed DevOps Pool to be created.

* `maximum_concurrency` - (Required) Defines how many resources can there be created at any given time.

* `name` - (Required) The name which should be used for this Managed DevOps Pool. Changing this forces a new Managed DevOps Pool to be created.

* `organization_profile` - (Required) An `organization_profile` block as defined below.

* `resource_group_name` - (Required) The name of the Resource Group where the Managed DevOps Pool should exist. Changing this forces a new Managed DevOps Pool to be created.

---

* `identity` - (Optional) An `identity` block as defined below.

* `tags` - (Optional) A mapping of tags which should be assigned to the Managed DevOps Pool.

---

An `agent_profile` block supports the following:

* `kind` - (Required) Defines the type of agent profile. Accepted values are: `Stateful` and `Stateless`.

> * `grace_period_time_span` - (Optional) How long should the machine be kept around after it ran a workload when there are no stand-by agents. The maximum is one week. This is needed when kind is `Stateful`.

Reviewer: what format is this in? seconds? iso duration?
```suggestion
* `grace_period_in_seconds` - (Optional) How long should the machine be kept around after it ran a workload when there are no stand-by agents. The maximum is one week. This is needed when kind is `Stateful`.
```
or 
```suggestion
* `grace_period_Duration` - (Optional) How long should the machine be kept around after it ran a workload when there are no stand-by agents in ISO8601 duration format, for example `P14D`. The maximum is one week. This is required when kind is `Stateful`.
```

Author: This is our public docs for our reference and want to keep it in the same way.
However, I have updated the description. please check.

https://learn.microsoft.com/en-us/azure/devops/managed-devops-pools/configure-scaling?view=azure-devops&tabs=azure-portal
![image](https://github.com/user-attachments/assets/6822928b-e899-49a2-8544-a936f4f9ba0c)

> * `max_agent_lifetime` - (Optional) How long should stateful machines be kept around. The maximum is one week. This is needed when kind is `Stateful`.

Reviewer: same as above

* `resource_predictions` - (Optional) A `resource_predictions` block as defined below.

* `resource_predictions_profile` - (Optional) One or more `resource_predictions_profile` blocks as defined below.

---

A `data_disks` block supports the following:

* `caching` - (Optional) The type of caching in a data disk. Accepted values are: `None`, `ReadOnly` and `ReadWrite`.

* `disk_size_gb` - (Optional) The initial disk size in gigabytes.

* `drive_letter` - (Optional) The drive letter for the empty data disk. If not specified, it will be the first available letter.

> * `storage_account_type` - (Optional) The storage Account type to be used for the data disk. If omitted, the default is "standard_lrs".

Reviewer 2: Added Possible values and Default value to `storage_account_type`

* `buffer` - (Optional) The percentage of the buffer to be allocated to this image.

* `resource_id` - (Optional) The resource id of the image.

* `well_known_image_name` - (Optional) The image to use from a well-known set of images made available to customers.

---

A `network_profile` block supports the following:

* `subnet_id` - (Required) The subnet id on which to put all machines created in the pool.

---

An `organization_profile` block supports the following:

* `kind` - (Required) Discriminator property for OrganizationProfile. Accepted value is`AzureDevOps` currently.

* `organizations` - (Required) One or more `organizations` blocks as defined below.

* `permission_profile` - (Optional) One or more `permission_profile` blocks as defined below.

---

An `organizations` block supports the following:

* `url` - (Required) The Azure DevOps organization URL in which the pool should be created.

* `parallelism` - (Optional) Specifies how many machines can be created at maximum in this organization out of the `maximum_concurrency` of the pool.

* `projects` - (Optional) List of projects in which the pool should be created.

---

An `os_profile` block supports the following:

> * `logon_type` - (Required) Determines how the service should be run. Accepted values are: `Interactive` and `Service`.

Reviewer 2: Updated to `Possible values are`

* `groups` - (Optional) Specifies a list of group email addresses.

* `users` - (Optional) Specifies a list of user email addresses.

---

A `resource_predictions` block supports the following:

* `days_data` - (Optional) A JSON string containing a list of maps, where each map represents a day of the week. Each map includes keys for start and end times, with corresponding values indicating the number of agents available during those times. For example, jsonencode([{},{"09:00:00": 1, "17:00:00": 0},{},{},{},{},{}]) specifies 1 standby agent available every Monday from 9:00 AM to 5:00 PM.

* `time_zone` - (Optional) Specifies the time zone for the predictions data to be provisioned at. Default is UTC.

---

A `resource_predictions_profile` block supports the following:

* `kind` - (Required) Determines how the stand-by scheme should be provided. Accepted values are: `Manual` and `Automatic`.

* `prediction_preference` - (Optional) Specifies the desired balance between cost and performance. Accepted values are: `MostCostEffective`, `MoreCostEffective`, `Balanced`, `MorePerformance`, and `BestPerformance`.

---

A `secrets_management_settings` block supports the following:

> * `certificate_store_location` - (Optional) Specified where to store certificates on the machine.

Reviewer 3: ```suggestion
* `certificate_store_location` - (Optional) Specifies where to store certificates on the machine.
```

* `certificate_store_name` - (Optional) Name of the certificate store to use on the machine. Accepted values are: 'My' and 'Root'.

* `key_exportable` - (Optional) Defines if the key of the certificates should be exportable.

* `observed_certificates` - (Optional) Specifies the list of certificates from Azure Key vault to install on all machines in the pool.

---

A `sku` block supports the following:

* `name` - (Required) The Azure SKU of the machines in the pool.

---

A `storage_profile` block supports the following:

* `data_disks` - (Optional) One or more `data_disks` blocks as defined above.

* `os_disk_storage_account_type` - (Optional) The storage account type of the OS disk. Accepted values are: `Premium`, `Standard` and `StandardSSD`.

## Attributes Reference

In addition to the Arguments listed above - the following Attributes are exported:

* `id` - The ID of the Managed DevOps Pool.

* `provisioning_state` - The status of the pool operation.

> * `type` - This is the resource type `Microsoft.DevOpsInfrastructure/pools`.

Reviewer 4: This isn't in the schema

}

resource "azurerm_dev_center_project" "example" {
  dev_center_id       = azurerm_dev_center.example.id
  location            = azurerm_resource_group.example.location
  name                = "example"
  resource_group_name = azurerm_resource_group.example.name
}

resource "azurerm_managed_devops_pool" "example" {
  name                           = "example"
  resource_group_name            = azurerm_resource_group.example.name
  location                       = azurerm_resource_group.example.location
  dev_center_project_resource_id = azurerm_dev_center_project.example.id
  maximum_concurrency            = 1

  azure_devops_organization_profile {
    organization {
      parallelism = 1
      url         = "https://dev.azure.com/example"
    }

    permission_profile_kind = "SpecificAccounts"

    administrator_accounts {
      groups = ["group1@example.com", "group2@example.com"]
      users  = ["user1@example.com", "user2@example.com"]
    }
  }

  stateless_agent_profile {
    automatic_resource_predictions_profile {
      prediction_preference = "Balanced"
    }
  }

  vmss_fabric_profile {
    sku {
      name = "Standard_D2ads_v5"
    }

    image {

>       resource_id = "/Subscriptions/00000000-0000-0000-0000-000000000000/Providers/Microsoft.Compute/Locations/australiaeast/publishers/canonical/artifacttypes/vmimage/offers/0001-com-ubuntu-server-focal/skus/20_04-lts-gen2/versions/latest"

Reviewer 2: Added an `azurerm_platform_image` data source, and reference it from there

      buffer      = "*"
    }

    storage_profile {
      os_disk_storage_account_type = "Standard"

      data_disk {
        disk_size_gb = 1
        drive_letter = "B"
      }
    }

    network_profile {

>       subnet_id = "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/group1/providers/Microsoft.Network/virtualNetworks/vnet1/subnets/subnet1"

Reviewer 5: Could we make this a reference to another `azurerm_subnet` resource defined above?

Reviewer 2: I have added a `azurerm_subnet` block and reference id from that resource. thanks!

>       subnet_id = "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/group1/providers/Microsoft.Network/virtualNetworks/vnet1/subnets/subnet1"

Reviewer 5: Can we reference a `azurerm_subnet` (created above) instead of using a hardcoded resource id?

Reviewer 2: I think these comments are duplicated for some reasons. But I have updated it as above

    }
  }
}
```

## Arguments Reference

The following arguments are supported:

> 

Reviewer 2: Updated

> * `dev_center_project_resource_id` - (Required) The ID of the Dev Center project.

Reviewer 5: Can we make sure the `name`, `resource_group_name` and `location` are positioned first, to be consistent with the other documents?

* `vmss_fabric_profile` - (Required) A `vmss_fabric_profile` block as defined below.

* `location` - (Required) The Azure Region where the Managed DevOps Pool should exist. Changing this forces a new Managed DevOps Pool to be created.

* `maximum_concurrency` - (Required) Defines how many resources can there be created at any given time.

* `name` - (Required) The name which should be used for this Managed DevOps Pool. Changing this forces a new Managed DevOps Pool to be created.

* `azure_devops_organization_profile` - (Required) An `azure_devops_organization_profile` block as defined below.

* `resource_group_name` - (Required) The name of the Resource Group where the Managed DevOps Pool should exist. Changing this forces a new Managed DevOps Pool to be created.

* `stateful_agent_profile` - (Optional) A `stateful_agent_profile` block as defined below. Exactly one of `stateful_agent_profile` or `stateless_agent_profile` must be specified.

> * `stateless_agent_profile` - (Optional) A `stateless_agent_profile` block as defined below. Exactly one of `stateful_agent_profile` or `stateless_agent_profile` must be specified.

Reviewer 5: Based on https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/reference-documentation-standards.md#warning-note, this shall be modified to:

```suggestion
* `stateful_agent_profile` - (Optional) A `stateful_agent_profile` block as defined below.

* `stateless_agent_profile` - (Optional) A `stateless_agent_profile` block as defined below.

~> **NOTE:** Exactly one of `stateful_agent_profile` or `stateless_agent_profile` must be specified.
```

> * `stateless_agent_profile` - (Optional) A `stateless_agent_profile` block as defined below. Exactly one of `stateful_agent_profile` or `stateless_agent_profile` must be specified.

Reviewer 5: Based on https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/reference-documentation-standards.md#warning-note, this shall be:

```suggestion
* `stateful_agent_profile` - (Optional) A `stateful_agent_profile` block as defined below.

* `stateless_agent_profile` - (Optional) A `stateless_agent_profile` block as defined below.

~> **NOTE:** Exactly one of `stateful_agent_profile` or `stateless_agent_profile` must be specified.
```

Reviewer 2: Updated as above

---

* `identity` - (Optional) An `identity` block as defined below.

* `tags` - (Optional) A mapping of tags which should be assigned to the Managed DevOps Pool.

---

A `stateful_agent_profile` block supports the following:

* `grace_period_time_span` - (Optional) Configures the amount of time an agent in a `stateful` pool waits for new jobs before shutting down after all current and queued jobs are complete. The format for Grace Period is `dd.hh:mm:ss`. Defaults to `00:00:00`.

* `max_agent_lifetime` - (Optional) Configures the maximum duration an agent in a `stateful` pool can run before it is shut down and discarded. The format for Max time to live for standby agents is `dd.hh:mm:ss`. Defaults to `7.00:00:00`.

* `manual_resource_predictions_profile` - (Optional) A `manual_resource_predictions_profile` block as defined below. Exactly one of `manual_resource_predictions_profile` or `automatic_resource_predictions_profile` may be specified.

> * `automatic_resource_predictions_profile` - (Optional) An `automatic_resource_predictions_profile` block as defined below. Exactly one of `manual_resource_predictions_profile` or `automatic_resource_predictions_profile` may be specified.

Reviewer 5: Same here.

---

A `stateless_agent_profile` block supports the following:

* `manual_resource_predictions_profile` - (Optional) A `manual_resource_predictions_profile` block as defined below. Exactly one of `manual_resource_predictions_profile` or `automatic_resource_predictions_profile` may be specified.

> * `automatic_resource_predictions_profile` - (Optional) An `automatic_resource_predictions_profile` block as defined below. Exactly one of `manual_resource_predictions_profile` or `automatic_resource_predictions_profile` may be specified.

Reviewer 5: Same here

> * `automatic_resource_predictions_profile` - (Optional) An `automatic_resource_predictions_profile` block as defined below. Exactly one of `manual_resource_predictions_profile` or `automatic_resource_predictions_profile` may be specified.

Reviewer 2: Updated as above

---

A `data_disk` block supports the following:

* `caching` - (Optional) The type of caching in a data disk. Possible values are `None`, `ReadOnly` and `ReadWrite`.

* `disk_size_gb` - (Optional) The initial disk size in gigabytes. Possible values are between `1` and `32767`.

* `drive_letter` - (Optional) The drive letter for the empty data disk. If not specified, it will be the first available letter.

* `storage_account_type` - (Optional) The Storage Account type to be used for the data disk. Possible values are `Premium_LRS`, `Premium_ZRS`, `Standard_LRS`, `StandardSSD_LRS` and `StandardSSD_ZRS`. Defaults to `Standard_LRS`.

---

A `vmss_fabric_profile` block supports the following:

* `image` - (Required) One or more `image` blocks as defined below.

> * `sku` - (Required) A `sku` block as defined below.

Reviewer 2: Good point! I have flattened it into `sku_name`. Thanks!

* `network_profile` - (Optional) A `network_profile` block as defined below.

* `os_profile` - (Optional) An `os_profile` block as defined below.

* `storage_profile` - (Optional) A `storage_profile` block as defined below.

---

An `identity` block supports the following:

* `type` - (Required) The type of managed service identity. Possible value is `UserAssigned`.

* `identity_ids` - (Required) Specifies a list User assigned managed identity Id's.

---

An `image` block supports the following:

* `aliases` - (Optional) List of aliases to reference the image by.

* `buffer` - (Optional) The percentage of the buffer to be allocated to this image.

* `resource_id` - (Optional) The resource id of the image.

* `well_known_image_name` - (Optional) The image to use from a well-known set of images made available to customers.

-> **Note:** More information about supported images can be found in [Microsoft Learn documentation](https://learn.microsoft.com/en-us/azure/devops/managed-devops-pools/configure-pool-settings?view=azure-devops&tabs=azure-portal#images)

> -> **Note:** Exactly one of `resource_id` and `well_known_image_name` are required per `image`

Reviewer 6: "An image can only have one or the other."

I think "and" should become "or" for clarity.

> -> **Note:** Exactly one of `resource_id` and `well_known_image_name` are required per `image`

Reviewer 5: ```suggestion
~> **Note:** Exactly one of `resource_id` and `well_known_image_name` are required per `image`
```
